src/order/height.lean

```diff
@@ -0,0 +1,276 @@
+
+lemma exists_chain_of_le_length {a : α} {l : list α} (hl : chain (<) a l) {n : ℕ}
+  (hn : n ≤ l.length) : ∃ l, chain (<) a l ∧ l.length = n :=
+begin
+  cases n,
+  { exact ⟨[], chain.nil, rfl⟩ },
+  cases e : l.drop n,
+  { exact (nat.lt_le_antisymm (nat.lt_iff_add_one_le.mpr hn) (drop_eq_nil_iff_le.mp e)).elim },
+  rw [← l.take_append_drop n, e] at hl,
+  refine ⟨_, (chain_split.mp hl).1, _⟩,
+  simpa [nat.succ_eq_add_one] using n.le_succ.trans hn
+end
+
+lemma exists_chain_of_le_chain_height {a : α} {n : ℕ} (hn : ↑n ≤ chain_height a) :
+  ∃ l, chain (<) a l ∧ l.length = n :=
+begin
+  by_cases ha : chain_height a = ⊤,
+  { obtain ⟨⟨l, hl⟩, hl'⟩ := enat.exists_le_of_supr_eq_top _ ha n,
+    apply exists_chain_of_le_length hl,
+    simpa using hl' },
+  { obtain ⟨⟨l, hl⟩, h : _ = chain_height a⟩ := enat.exists_eq_supr_of_ne_top _ ha,
+    apply exists_chain_of_le_length hl,
+    rw ← h at hn,
+    simpa using hn }
+end
+
+lemma exists_chain_of_le_sup_chain_height {n : ℕ} (hn : ↑n ≤ sup_chain_length α) :
+  ∃ l : list α, chain' (<) l ∧ l.length = n :=
+begin
+  cases n,
+  { exact ⟨[], by simp⟩ },
+  by_cases ha : sup_chain_length α = ⊤,
+  { obtain ⟨⟨l, hl⟩, hl'⟩ := enat.exists_le_of_supr_eq_top _ ha (n+1),
+    cases l,
+    { simp only [list.length, enat.coe_le_coe, nat.succ_ne_zero, nonpos_iff_eq_zero] at hl',
+      exact hl'.elim },
+    dsimp at *,
+    rw [enat.coe_le_coe, add_le_add_iff_right] at hl',
+    obtain ⟨l', h, h'⟩ := exists_chain_of_le_length hl hl',
+    rw ← h',
+    refine ⟨l_hd :: l', h, rfl⟩ },
+  casesI is_empty_or_nonempty {l : list α // chain' has_lt.lt l} with H H,
+  { rw [sup_chain_length, supr, set.range_eq_empty, Sup_empty, le_bot_iff] at hn,
+    exact (n.succ_ne_zero (enat.coe_inj.mp hn)).elim },
+  { obtain ⟨⟨l, hl⟩, h : ↑l.length = sup_chain_length α⟩ := enat.exists_eq_supr_of_ne_top _ ha,
+    cases l,
+    { simp only [← h, list.length, enat.coe_le_coe, nat.succ_ne_zero, nonpos_iff_eq_zero] at hn,
+      exact hn.elim },
+    have : n ≤ l_tl.length,
+    { rw [← h, enat.coe_le_coe] at hn, exact (add_le_add_iff_right 1).mp hn },
+    obtain ⟨l', h, h'⟩ := exists_chain_of_le_length hl this,
+    rw ← h',
+    refine ⟨l_hd :: l', h, rfl⟩ }
+end
+
+lemma enat.ne_zero_iff_one_le {n : enat} : n ≠ 0 ↔ 1 ≤ n :=
+begin
+  change ¬ n = ⊥ ↔ _,
+  rw [eq_bot_iff, ← lt_iff_not_le],
+  convert (enat.add_one_le_iff_lt (show (0 : enat) ≠ ⊤, from enat.coe_ne_top 0)).symm,
+  exact (zero_add 1).symm
+end
+
+lemma chain_height_eq_zero_iff (a : α) : chain_height a = 0 ↔ ∀ b, ¬ a < b :=
+begin
+  rw ← not_iff_not,
+  push_neg,
+  rw enat.ne_zero_iff_one_le,
+  split,
+  { intro h,
+    obtain ⟨l, hl, hl'⟩ := exists_chain_of_le_chain_height h,
+    obtain ⟨b, rfl⟩ := list.length_eq_one.mp hl',
+    cases hl,
+    exact ⟨b, by assumption⟩ },
+  { rintro ⟨b, r⟩,
+    refine le_trans _ (le_supr _ ⟨[b], by simpa⟩),
+    simp }
+end
+
+lemma sup_chain_length_eq_zero_iff  : sup_chain_length α = 0 ↔ is_empty α :=
```

Comment:
You should also have a `simp` lemma stating that when `is_empty α`, then `sup_chain_length α = 0`.